iup/__init__.py

```diff
@@ -173,45 +199,37 @@ def insert_rollout(
         """
         frame = self
 
-        for r in rollout:
-            if group_cols is not None:
-                rollout_rows = (
-                    frame.select(pl.col(v) for v in group_cols)
-                    .unique()
-                    .with_columns(date=r, estimate=0.0)
-                )
-            else:
-                rollout_rows = pl.DataFrame({"date": r, "estimate": 0.0})
-
-            frame = frame.vstack(rollout_rows.select(frame.columns)).sort("date")
-
-        return frame
-
-
-def extract_group_names(
-    group_cols: List[List[str]],
-) -> tuple | None:
-    """
-    Note the column names for grouping factors across data sets.
-
-    Parameters
-    group_cols: [[str,]]
-        List of lists of column names for grouping factors, for each data set
-
-    Returns
-        (str,)
-        The column names of grouping factors common to all data sets
+        if group_cols is not None:
+            rollout_rows = (
+                frame.select(group_cols)
+                .unique()
+                .join(pl.DataFrame({"time_end": rollout}), how="cross")
+                .with_columns(estimate=0.0)
+            )
+            group_cols = group_cols + ["season"]
+        else:
+            rollout_rows = pl.DataFrame({"time_end": rollout, "estimate": 0.0})
+            group_cols = ["season"]
 
-    Details
-    Before returning a single tuple of the desired column names,
-    check that they are identical for every data set.
-    """
+        frame = frame.vstack(rollout_rows.select(frame.columns)).sort("time_end")
```

Comment:
```suggestion
        frame = rollout_rows.select(frame.columns).sort("time_end")
```
you don't need the vstack, since now you only have a single object!

Reply:
I think we do need the `vstack` because `rollout_rows` contains _only_ rows with 0 uptake on the rollout dates. All the other non-zero data is in `frame`. So we must stack the two together to get the full data set with rollout included.

Reply:
got it